Use Memcached CAS to remember urls if available.

YPCrumble opened the pull request:
Fixes #7.

@peterbe thanks for considering this PR! I've done some research into the issue with Memcached race conditions and believe this will solve the problem using Memcached CAS ("check and set"). It's simpler than storing `remembered_urls` in the db.

I've implemented such that the code tries using Memcached CAS and falls back to the current functionality in the event of an `AttributeError` that shows that `cache.cas` is unavailable.

Please let me know if you have any thoughts or improvements. Thanks again for maintaining this great repo!

peterbe commented:
Hi @YPCrumble 
To be honest, I'm struggling to review this. 
I haven't used this package in years. Does this PR solve your problem? 
Was upgrading to `django-redis` not an option?

I see that unfortunately this project doesn't have any automated tests or release tools. Do the tests pass as of this PR?

What I don't like is that exceptions are swallowed. Doing `LOGGER.warn(...)` is to not deal with the error. Nobody looks at log files except in extreme cases. The only time I think it's OK to swallow exceptions is if you have no hope in hell to more carefully test if it's going to work. E.g
```python
try:
  some_network_get(...)
except HTTPConnectionError:
  print("Unable to connect due to network error")
```

So if you can rewrite your `except:` blocks to not log things unless you absolutely have to for the CAS case. Then I'll merge it.

YPCrumble commented:
@peterbe agree it's a little odd to swallow the Exception, but I've added some comments that hopefully explain _why_ we're swallowing the exception.

Here's a little more detail on my reasoning as well:

Memcached CAS is accessible only via an undocumented and private API in Django and as such is subject to change in the future (though I've made a little headway in seeing if I can add CAS functionality to Django core here: https://groups.google.com/g/django-developers/c/GVb8--s1g6g/m/PEJqsMJqEAAJ).

Because the underlying API is private and subject to change, and because it isn't really "supported" by Django, there are a number of exceptions that can be raised when trying to access the API, and these exceptions could change in the future. The exceptions are also different if we're using pymemcached or pylibmc, and I didn't want to attempt to specify all possible exceptions.

Because of this, essentially the way I've rewritten `remember_url` is to do this:

```
try:
    If CAS is available, this means we're using Memcached. Use CAS to save the remembered_url to the REMEMBERED_URLS_KEY and return. This solves the thread safety issue.
except Exception:
    CAS is not available either because we're on Redis or because something unexpected has happened - fall back to the existing remember_url functionality.

...existing remember_url functionality continues here and runs if we're using Redis or something unexpected has happened...
```

The above code handles redis gracefully because `cache._cache.cas` will raise an AttributeError and we simply move on to the existing remember_url functionality that works fine with Redis. It's I suppose an "Ask for forgiveness" strategy. If you prefer I could check for `hasattr(self.cache._cache, 'cas')` before trying `remember_url_cas`, but I thought handling the `AttributeError` does the same thing in a simpler manner.

Using Redis is an option but for my use case Memcached has been working really well and this is the only issue I've ever had - so making this change is likely a smaller lift than switching the entire caching infrastructure to Redis.

Let me know your thoughts on the above?

YPCrumble commented:
@peterbe also - the logs would only ever be used if someone is trying to resolve the issue raised in #7. If either of the two loggers gets hit, the code works exactly the same as it does currently, but the errors (which I'll change to warning now) simply alert the user that they may run into #7 in a high traffic environment.

peterbe commented:
You might need to merge in latest `master` now.

YPCrumble commented:
@peterbe just merged in latest `master` - thank you! Would you mind making a release once this is merged in? Or let me know if that's possible for me to do and I'm happy to make a PR!

peterbe reviewed (changes requested):
I think this is the wrong approach. Sorry. But I love that you've found a solution to the original problem. 
What I don't like is that it's very "oddly specific" and it swallows too many errors. 

Thing is if you use this lib with Memcache instead of Redis, the only acceptable way is to use your solution. It's also a bit weird that logging is suddenly introduced as part of the thread-safety problem. I would love to see logging added because I think it can help and if configured correctly, using the `logging` module, it's not a bad thing to have. What you should do is **first figure out if Memcached is the backend** and if so just use your `remember_url_cas()` without much "excuses or options". I was thinking something like this, inside `remember_url()`:

```python
def remember_url(self, request, cache_key, timeout):
    if self._is_memcache_backend():
        return self.remember_url_cas(url, cache_key)
    # If it's not Memcache, don't try to solve the thread-safety problem. 
    # TODO: Implement a solution for Redis 
    # See https://github.com/peterbe/django-fancy-cache/issues/7
    #
    # This is the default basic implementation which is not thread-safe.
    
     ... the old code ...

```

Regarding `_is_memcache_backend` perhaps we can do something like:
```python
hasattr(cache, 'make_key')
# or...
'memcached' settings.CACHES['default']['BACKEND']
```
I honestly don't know what is the best way to find out if `cache` is a proxy for Memcache. Experiment. 

And lastly, remove all `try:` and `except ...:` that aren't absolutely necessary.

peterbe commented:
I see that you squashed the commits so I can't tell what has changed. If you're able, please avoid rebase and only merge in `master` periodically.

YPCrumble commented:
@peterbe will do! I'm at a company that requires squashing which like you I prefer not to do, so I'll add commits in the future :D. 

Just made an update as I'm currently user-testing this on my website via a fork. I actually removed `make_key` which wasn't necessary as fancy-cache caches the REMEMBERED_URLS directly under `fancy-cache`.

Will update this PR soon once I've made sure it works and have addressed your very helpful review suggestions.

peterbe commented:
> I'm at a company that requires squashing which like you I prefer not to do, so I'll add commits in the future :D.

What we do, at my work, is we never rebase anything ever (unless in extreme cases or to avoid embarrassing commit messages), but when we merge the PR we *always* use "Squash and merge". Then you get the best of both worlds. Clean commit history but full audit in the PR work.

YPCrumble commented:
@peterbe thanks for your thoughtful review!

I've made an update that does a few things:

1. It uses your helpful idea of a private method to determine whether CAS is available
2. It removes the try...except blocks. After user-testing this for a couple months on my server it seems to work fine without any unusual exceptions, so I agree it makes sense to remove these. One note is that this is a breaking change for Memcached users - they will now have to enable CAS. Per your comment above, the remember_url function is really broken anyway for Memcached, so I agree a breaking change makes sense here :D.
3. Adds documentation warning Memcached users that CAS is now required for use with Memcached.

I apologize in advance also - I added three commits and realized there was a stupid bug in one of them, and so I force-pushed the three commits again. I don't _think_ that breaks your workflow but I did it so fast without thinking and didn't mean to do that, mea culpa!

Thanks again for your reviews and thoughtfulness about these updates!

YPCrumble commented:
@peterbe I updated this to use a settings flag per your comments in #36 - mind taking another look?

Thanks again for your reviews and patience with this!

peterbe reviewed (approved):
Cool. So it's totally optional now. By default. Seems safe and sane.

peterbe commented:
I can't remember how I deploy to PyPI on this.
